\documentclass[a4paper]{article}
\usepackage{hyperref,fullpage}

\title{BCGES short courses, session 1 \newline RNA-Seq}
\date{}
\author{}

\begin{document}
%\SweaveOpts{concordance=TRUE}
%knit('practical.Rnw'); system('pdflatex practical')

\maketitle


<<setup, results = 'hide', echo = FALSE>>=
forStudents <- TRUE
if (!forStudents) show <- 'markup' else show <- 'hide'
@


\section{\texttt{samtools} over the web to download BAM files}

A powerful feature of BAM/SAM and \texttt{samtools} is the ability to use the \texttt{view} feature with indexes over the web, to only pull the chunks of interest.
I will provide an example of this using the 1,000 Genomes data.
To do that, one should first download the list of whole genome alignments from the 1,000 Genomes to identify where the files of interest are located on the server.
<<wget, eval = FALSE, engine = 'bash'>>=
wget -O 1KG_alignment.index ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/alignment.index
@ 


%%%%%%%%%%%%%%%%%%%%%%
\section{Read depth analysis to detect CNVs}
<<exomeDepth>>=
library(ExomeDepth)
@

%%%%%%%%%%%%%%%%%%%%%%
\section{Use Rsamtools to identify reads characteristic of a deletion}

Another strategy to detect CNVs consists of picking up reads that show an unusual pattern.
For example, two pairs further apart than they normally should is potentially flagging a deletion.
In this exercice I proposed to look at a well described 20 kb CNV located near the {\it IRGM} gene.
We can start by looking at the location of the variant using the data from the Conrad et al paper, which are loaded into ExomeDepth.

<<IRGM, eval = TRUE>>=
data(Conrad.hg19)
IRGM <- Conrad.hg19.common.CNVs[ seqnames(Conrad.hg19.common.CNVs) == 5 &
                                start(Conrad.hg19.common.CNVs) > 150200000
                                & end(Conrad.hg19.common.CNVs) < 150240000,]
write.table(x = as.data.frame(IRGM),
            row.names = FALSE, sep = '\t', quote = FALSE,
            file = 'IRGM_common.tab')
@ 


We now want to look for individuals that carry this variant. 
Low coverage whole genome data for the 1KG sample HG00123 will do the job.

\noindent
{\bf Exercise:} Identify the BAM file for mapped reads for HG00123 and download the slice 5:150200000-150225000 using \texttt{samtools view}.
How would you filter pairs of reads that span over the deletion in this sample?


<<download, eval = FALSE, engine = 'bash'>>=
samtools view -b ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data/HG00123/alignment/HG00123.mapped.ILLUMINA.bwa.GBR.low_coverage.20120522.bam 5:150200000-150225000 > file.bam
samtools view file.bam  | awk '{if (($9 > 15000) && ($9 < 25000)) print}' | less -S
@

{\bf (Optional) Exercise:} Can you write a R script (using Rsamtools) to identify these reads?







\end{document}
