\documentclass[a4paper]{article}
\usepackage{hyperref,fullpage}

\title{BCGES short courses, session 1 \newline RNA-Seq}
\date{}
\author{}

\begin{document}
%\SweaveOpts{concordance=TRUE}
%knit('practical.Rnw'); system('pdflatex practical')

\maketitle


<<setup, results = 'hide', echo = FALSE>>=
forStudents <- TRUE
if (!forStudents) show <- 'markup' else show <- 'hide'
@



%%%%%%%%%%%%%%%%%%%%%%
\section{Read depth analysis to detect CNVs}



%%%%%%%%%%%%%%%%%%%%%%
\section{Use Rsamtools to identify reads characteristic of a deletion}

Another strategy to detect CNVs consists of picking up reads that show an unusual pattern.
For example, two pairs further apart than they normally should is potentially flagging a deletion.



<<get.odd.reads>>=
library(Rsamtools)
library(GenomicRanges)

which <- GRanges(seqnames=Rle('20'),
                 IRanges(start = 2854142, end = 3019315))

what <- c("rname", "strand", "pos", "qwidth", "seq")
param <- ScanBamParam(which=which, what=what)

bam <- scanBam(file = '../data/BAM_files/HG00096.mapped.ILLUMINA.bwa.GBR.exome.20120522.bam', param=param)
@ 





\end{document}
